hooks/context_change.py

```diff
@@ -0,0 +1,40 @@
+# Copyright (c) 2016 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+"""
+Context change hook.
+"""
+
+from tank import get_hook_baseclass
+
+
+class ContextChangeHook(get_hook_baseclass()):
+    """
+    Hook that gets executed every time there is a context change, once before
+    the change and once after.
+    """
+
+    def pre_context_change(self, current_context, next_context):
+        """
+        Called before the context change happens.
+
+        :param current_context: The current context.
+        :param next_context: The context being switching to.
+        """
+        pass
```

Comment:
It might be an interesting feature to implement at some point, but I'm concerned about the impact this might have on apps that switch the current context, namely workfiles 1&2.

For example, let's say you hit the New button. The folder creation runs with the context as selected by workfiles. Then, the context changes, which triggers the hooks. If the hooks updates the context by switching some value inside it, it means now that the wrong folders have been created. Maybe the hook that you are suggesting should be at the app level after the selection is made and before a file is {created,opened,saved}?

In any case, since the current hook returns None, nothing prevents us from amending the signature in the future to allow it to return a different context, which we could support in a backwards compatible manner by assuming None means "use the context passed in originally".

What do you think?